# Mini Progetto Intelligenza Artificiale - Riconoscimento cifre manoscritte

**Nome:** Giulio    
**Cognome:** Bottacin    
**Matricola:** 2042340    
**Data consegna:** 5/6/2025  

## Obiettivo

In questo progetto esploreremo il riconoscimento di cifre manoscritte utilizzando il dataset MNIST, implementando simulazioni per studiare come diversi fattori influenzano le prestazioni dei modelli di deep learning. Analizzeremo in particolare l'impatto degli iperparametri, la robustezza al rumore e l'effetto della quantità di dati di training.

## Importazione delle librerie necessarie

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import sklearn.metrics as metrics
from sklearn.neural_network import MLPClassifier
from torchvision.datasets import MNIST, FashionMNIST
from tensorflow import keras
import tensorflow as tf
import pandas as pd
import time
import warnings
warnings.filterwarnings('ignore')

# Configurazione per riproducibilità
np.random.seed(42)
tf.random.set_seed(42)
plt.rcParams['figure.figsize'] = (10, 6)

## Caricamento e preparazione del dataset MNIST

In [ ]:
# Caricamento dataset MNIST
print("Caricamento dataset MNIST...")
mnist_tr = MNIST(root="./data", train=True, download=True)
mnist_te = MNIST(root="./data", train=False, download=True)

Caricamento dataset MNIST...


In [ ]:
# Conversione in array numpy
mnist_tr_data, mnist_tr_labels = mnist_tr.data.numpy(), mnist_tr.targets.numpy()
mnist_te_data, mnist_te_labels = mnist_te.data.numpy(), mnist_te.targets.numpy()

# Preprocessing per MLP (vettorizzazione e normalizzazione)
x_tr = mnist_tr_data.reshape(60000, 28 * 28) / 255.0
x_te = mnist_te_data.reshape(10000, 28 * 28) / 255.0

# Preprocessing per CNN (mantenendo formato 2D)
x_tr_conv = x_tr.reshape(-1, 28, 28, 1)
x_te_conv = x_te.reshape(-1, 28, 28, 1)

print(f"Dataset caricato: {x_tr.shape[0]} esempi di training, {x_te.shape[0]} esempi di test")
print(f"Forma dati MLP: {x_tr.shape}")
print(f"Forma dati CNN: {x_tr_conv.shape}")

# Visualizzazione esempi del dataset
fig, axes = plt.subplots(2, 5, figsize=(12, 6))
fig.suptitle('Dataset MNIST - Esempi per Cifra', fontsize=14)

for digit in range(10):
    idx = np.where(mnist_tr_labels == digit)[0][0]
    ax = axes[digit//5, digit%5]
    ax.imshow(mnist_tr_data[idx], cmap='gray')
    ax.set_title(f'Cifra {digit}')
    ax.axis('off')

plt.tight_layout()
plt.show()

## Punto A: Effetto degli iperparametri sulle prestazioni

Analizziamo sistematicamente come variano le prestazioni dei modelli MLP e CNN al variare degli iperparametri chiave.  
Confronteremo 18 configurazioni MLP e 6 configurazioni CNN per un totale di 24 esperimenti mirati.

### Configurazione esperimenti sistematici

***MLP (18 esperimenti):***
- **Neuroni per strato**: *50, 100, 250* per testare la copertura da reti piccole a medio-grandi
- **Numero layers**: *1 vs 2* strati nascosti per fare il confronto profondità vs larghezza
- **Learning rate**: *0.001, 0.01, 0.1*

***CNN (6 esperimenti):***
- **Filtri**: *32*, standard per MNIST, computazionalmente efficiente
- **Architettura**: *baseline vs extended* per fare il confronto sulla complessità
- **Learning rate**: *0.001, 0.01, 0.1*

Per entrambi i modelli si è scelto di utilizzare il solver **Adam**, ormai standard e più performante di SDG.  
Si è volutamente scelto di eseguire meno esperimenti sulle CNN in quanto richiedono tempi molto più lunghi di training rispetto alle MLP.

#### Scelta dei parametri di training

***MLP:***
- *max_iter = 100* è sufficiente per convergenza su MNIST basato su cifre manoscritte. 
- *early_stopping = True*, previene l'overfitting essenziale quando sono presenti molti parametri.
- *validation_fraction = 0.1*, split standard 90/10.
- *tol = 0.001* è una precisione ragionevole per classificazione.
- *n_iter_no_change = 10* è un livello di pazienza adeguata per permettere oscillazioni temporanee.

***CNN:*** 
- *epochs = 20* valore di compromesso per bilanciare velocità e convergenza, il valore è più basso delle MLP perchè le CNN tipicamente convergono più velocemente.
- *batch_size = 128*, trade-off memoria/velocità ottimale per dataset size.
- *validation_split = 0.1*, coerente con le scelte di MLP.
- *patience = 5*, le CNN sono meno soggette a oscillazioni quindi è stato scelto un livello di pazienza minore.
- *min_delta = 0.001*, scelta la stessa precisione degli MLP per comparabilità diretta.

Questa configurazione permette un confronto sistematico e bilanciato tra i due tipi di architetture.

#### Funzioni helper per stampe risultati

In [ ]:
def stampa_header_esperimento(num_esp, totale, tipo_modello, config):
    print(f"\n[{num_esp:2d}/{totale}] {tipo_modello}: {config}")
    print("-" * 50)

def stampa_risultati_esperimento(risultati):
    print(f"Accuracy Training: {risultati['train_accuracy']:.4f} | Accuracy Test: {risultati['test_accuracy']:.4f}")
    print(f"Tempo: {risultati['training_time']:6.1f}s | Iterazioni: {risultati['iterations']:3d}")
    print(f"Overfitting: {risultati['overfitting']:+.4f}")

#### Esperimenti MLP (16 configurazioni)

In [ ]:
neuroni_lista = [50, 100, 250] # numero di neuroni per strato
strati_lista = [1, 2]  # numero di strati nascosti
learning_rates = [0.001, 0.01, 0.1] # learning rates 

risultati_mlp = []
contatore_esperimenti = 0
esperimenti_totali = len(neuroni_lista) * len(strati_lista) * len(learning_rates)

print("INIZIO ESPERIMENTI MLP")
print("=" * 60)

for neuroni in neuroni_lista:
    for n_strati in strati_lista:
        for lr in learning_rates:
            contatore_esperimenti += 1
            
            # Configurazione architettura
            if n_strati == 1:
                strati_nascosti = (neuroni,)
                nome_config = f"{neuroni}n_1S_lr{lr}"
            else:
                strati_nascosti = (neuroni, neuroni)
                nome_config = f"{neuroni}n_2S_lr{lr}"
            
            stampa_header_esperimento(contatore_esperimenti, esperimenti_totali, "MLP", nome_config)
            
            # Training MLP
            mlp = MLPClassifier(
                hidden_layer_sizes=strati_nascosti,
                learning_rate_init=lr,
                max_iter=100,
                early_stopping=True,
                validation_fraction=0.1,
                tol=0.001,
                n_iter_no_change=10,
                random_state=42
            )
            
            tempo_inizio = time.time()
            mlp.fit(x_tr, mnist_tr_labels)
            tempo_training = time.time() - tempo_inizio
            
            acc_train = mlp.score(x_tr, mnist_tr_labels)
            acc_test = mlp.score(x_te, mnist_te_labels)
            
            risultati = {
                'tipo_modello': 'MLP',
                'nome_config': nome_config,
                'neuroni': neuroni,
                'n_strati': n_strati,
                'learning_rate': lr,
                'strati_nascosti': strati_nascosti,
                'train_accuracy': acc_train,
                'test_accuracy': acc_test,
                'overfitting': acc_train - acc_test,
                'training_time': tempo_training,
                'iterations': mlp.n_iter_,
                'loss_curve': mlp.loss_curve_ if hasattr(mlp, 'loss_curve_') else [],
                'parametri_totali': sum([layer.size for layer in mlp.coefs_]) + sum([layer.size for layer in mlp.intercepts_])
            }
            
            risultati_mlp.append(risultati)
            stampa_risultati_esperimento(risultati)

print(f"\nESPERIMENTI MLP COMPLETATI")

INIZIO ESPERIMENTI MLP

[ 1/18] MLP: 50n_1S_lr0.001
--------------------------------------------------
Acc Training: 0.9891 | Acc Test: 0.9707
Tempo:    9.0s | Iterazioni:  24
Overfitting: +0.0184

[ 2/18] MLP: 50n_1S_lr0.01
--------------------------------------------------
Acc Training: 0.9844 | Acc Test: 0.9697
Tempo:    5.2s | Iterazioni:  17
Overfitting: +0.0147

[ 3/18] MLP: 50n_1S_lr0.1
--------------------------------------------------
Acc Training: 0.9202 | Acc Test: 0.9123
Tempo:    5.8s | Iterazioni:  20
Overfitting: +0.0079

[ 4/18] MLP: 50n_2S_lr0.001
--------------------------------------------------
Acc Training: 0.9905 | Acc Test: 0.9729
Tempo:    9.4s | Iterazioni:  27
Overfitting: +0.0176

[ 5/18] MLP: 50n_2S_lr0.01
--------------------------------------------------
Acc Training: 0.9863 | Acc Test: 0.9695
Tempo:    6.3s | Iterazioni:  19
Overfitting: +0.0168

[ 6/18] MLP: 50n_2S_lr0.1
--------------------------------------------------
Acc Training: 0.8471 | Acc Test: 

#### Funzioni helper per esperimenti CNN

In [ ]:
def crea_modello_cnn(tipo_architettura, learning_rate):
    model = keras.Sequential()
    
    if tipo_architettura == 'baseline':
        # Architettura baseline
        model.add(keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)))
        model.add(keras.layers.Flatten())
        model.add(keras.layers.Dense(50, activation='relu'))
        
    elif tipo_architettura == 'extended':
        # Architettura estesa con pooling e più strati
        model.add(keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)))
        model.add(keras.layers.MaxPooling2D(2,2))
        model.add(keras.layers.Conv2D(64, (3,3), activation='relu'))
        model.add(keras.layers.Flatten())
        model.add(keras.layers.Dense(100, activation='relu'))
    
    model.add(keras.layers.Dense(10, activation='softmax'))
    
    # Configurazione optimizer
    optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
    
    model.compile(
        optimizer=optimizer,
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    
    return model

#### Esperimenti CNN

In [ ]:
architetture = ['baseline', 'extended']
learning_rates_cnn = [0.001, 0.01, 0.1]

risultati_cnn = []
contatore_esperimenti_cnn = 0
esperimenti_totali_cnn = len(architetture) * len(learning_rates_cnn)

print("\n\nINIZIO ESPERIMENTI CNN")
print("=" * 60)

for arch in architetture:
    for lr in learning_rates_cnn:
        contatore_esperimenti_cnn += 1
        nome_config = f"CNN_{arch}_lr{lr}"
        
        stampa_header_esperimento(contatore_esperimenti_cnn, esperimenti_totali_cnn, "CNN", nome_config)
        
        # Creazione e training CNN
        model = crea_modello_cnn(arch, lr)
        
        # Early stopping callback
        early_stopping = keras.callbacks.EarlyStopping(
            patience=5,
            min_delta=0.001,
            restore_best_weights=True,
            verbose=0
        )
        
        tempo_inizio = time.time()
        history = model.fit(
            x_tr_conv, mnist_tr_labels,
            validation_split=0.1,
            epochs=20,
            batch_size=128,
            callbacks=[early_stopping],
            verbose=0
        )
        tempo_training = time.time() - tempo_inizio
        
        # Valutazione
        train_loss, acc_train = model.evaluate(x_tr_conv, mnist_tr_labels, verbose=0)
        test_loss, acc_test = model.evaluate(x_te_conv, mnist_te_labels, verbose=0)
        
        risultati = {
            'tipo_modello': 'CNN',
            'nome_config': nome_config,
            'architettura': arch,
            'learning_rate': lr,
            'train_accuracy': acc_train,
            'test_accuracy': acc_test,
            'overfitting': acc_train - acc_test,
            'training_time': tempo_training,
            'iterations': len(history.history['loss']),
            'loss_curve': history.history['loss'],
            'val_loss_curve': history.history['val_loss'],
            'parametri_totali': model.count_params()
        }
        
        risultati_cnn.append(risultati)
        stampa_risultati_esperimento(risultati)

print(f"\nESPERIMENTI CNN COMPLETATI")



INIZIO ESPERIMENTI CNN

[ 1/6] CNN: CNN_baseline_lr0.001
--------------------------------------------------
Acc Training: 0.9914 | Acc Test: 0.9805
Tempo:   68.4s | Iterazioni:   8
Overfitting: +0.0109

[ 2/6] CNN: CNN_baseline_lr0.01
--------------------------------------------------
Acc Training: 0.9890 | Acc Test: 0.9765
Tempo:   61.4s | Iterazioni:   7
Overfitting: +0.0125

[ 3/6] CNN: CNN_baseline_lr0.1
--------------------------------------------------
Acc Training: 0.1022 | Acc Test: 0.1010
Tempo:   55.1s | Iterazioni:   6
Overfitting: +0.0012

[ 4/6] CNN: CNN_extended_lr0.001
--------------------------------------------------
Acc Training: 0.9920 | Acc Test: 0.9893
Tempo:  107.7s | Iterazioni:   7
Overfitting: +0.0027

[ 5/6] CNN: CNN_extended_lr0.01
--------------------------------------------------
Acc Training: 0.9876 | Acc Test: 0.9804
Tempo:  108.9s | Iterazioni:   7
Overfitting: +0.0072

[ 6/6] CNN: CNN_extended_lr0.1
--------------------------------------------------
A

In [ ]:
# Combinazione risultati per analisi
tutti_risultati = risultati_mlp + risultati_cnn
df_risultati = pd.DataFrame(tutti_risultati)

### Grafico 1: Effetto del Learning Rate sulle prestazioni MLP

Questo grafico analizza l'impatto critico del learning rate sulla convergenza e stabilità del training per le reti MLP. Il learning rate controlla la dimensione dei passi durante l'ottimizzazione: valori troppo alti causano instabilità e divergenza, mentre valori troppo bassi rallentano eccessivamente la convergenza. L'analisi delle curve di loss e delle accuratezze finali permette di identificare il range ottimale per il dataset MNIST.

In [ ]:
# Preparazione dati per analisi learning rate
dati_lr_001 = [r for r in risultati_mlp if r['learning_rate'] == 0.001]
dati_lr_01 = [r for r in risultati_mlp if r['learning_rate'] == 0.01]
dati_lr_1 = [r for r in risultati_mlp if r['learning_rate'] == 0.1]

# Calcolo medie per ogni learning rate
acc_lr_001 = np.mean([r['test_accuracy'] for r in dati_lr_001])
acc_lr_01 = np.mean([r['test_accuracy'] for r in dati_lr_01])
acc_lr_1 = np.mean([r['test_accuracy'] for r in dati_lr_1])

tempo_lr_001 = np.mean([r['training_time'] for r in dati_lr_001])
tempo_lr_01 = np.mean([r['training_time'] for r in dati_lr_01])
tempo_lr_1 = np.mean([r['training_time'] for r in dati_lr_1])

# Visualizzazione curve di loss rappresentative
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6))

# Subplot 1: Curve di loss
for i, (dati_lr, colore, etichetta) in enumerate([(dati_lr_001, 'green', 'LR=0.001'), 
                                           (dati_lr_01, 'blue', 'LR=0.01'), 
                                           (dati_lr_1, 'red', 'LR=0.1')]):
    if dati_lr and dati_lr[0]['loss_curve']:
        curva_loss = dati_lr[0]['loss_curve']  # Primo esempio rappresentativo
        ax1.plot(range(len(curva_loss)), curva_loss, color=colore, linewidth=2, label=etichetta)

ax1.set_xlabel('Iterazioni')
ax1.set_ylabel('Loss')
ax1.set_title('Pattern di Convergenza per Learning Rate')
ax1.legend()
ax1.grid(True, alpha=0.3)

# Subplot 2: Accuratezza vs Learning Rate
learning_rates_plot = [0.001, 0.01, 0.1]
accuratezze = [acc_lr_001, acc_lr_01, acc_lr_1]
colori = ['green', 'blue', 'red']

bars = ax2.bar(range(len(learning_rates_plot)), accuratezze, color=colori, alpha=0.7)
ax2.set_xlabel('Learning Rate')
ax2.set_ylabel('Accuratezza Test Media')
ax2.set_title('Accuratezza Test per Learning Rate')
ax2.set_xticks(range(len(learning_rates_plot)))
ax2.set_xticklabels(['0.001', '0.01', '0.1'])
ax2.grid(True, alpha=0.3)

# Annotazioni valori
for bar, acc in zip(bars, accuratezze):
    height = bar.get_height()
    ax2.annotate(f'{acc:.3f}', xy=(bar.get_x() + bar.get_width()/2, height),
                xytext=(0, 3), textcoords="offset points", ha='center', va='bottom')

plt.tight_layout()
plt.show()

print("DATI ANALISI LEARNING RATE:")
print(f"LR=0.001: Accuratezza={acc_lr_001:.4f}, Tempo={tempo_lr_001:.1f}s")
print(f"LR=0.01:  Accuratezza={acc_lr_01:.4f}, Tempo={tempo_lr_01:.1f}s") 
print(f"LR=0.1:   Accuratezza={acc_lr_1:.4f}, Tempo={tempo_lr_1:.1f}s")

#### Discussione del grafico e riflessione sui risultati
I risultati mostrano chiaramente l'effetto critico del learning rate sulle prestazioni: 
- **LR = 0.001** raggiunge la migliore accuratezza media (97.65%) con convergenza stabile ma lenta
- **LR = 0.01** mantiene prestazioni competitive (97.32%) con convergenza più rapida rappresentando il miglior compromesso velocità-accuratezza
- **LR = 0.1** causa un drammatico crollo delle prestazioni (86.12%) indicando instabilità nell'ottimizzazione e possibili oscillazioni eccessive. 

Dal punto di vista tecnico, learning rate alti causano passi troppo grandi che fanno "saltare" oltre i minimi locali, mentre valori troppo bassi intrappolano l'ottimizzazione in plateau prolungati. 

Per applicazioni pratiche, si raccomanda l'uso di LR=0.01 come punto di partenza per MLP su MNIST, con possibile fine-tuning verso 0.001 se il tempo di training non è critico e si desidera massimizzare l'accuratezza finale.

### Grafico 2: Confronto Completo delle Architetture (Training vs Test)
Questo grafico presenta un confronto esaustivo di tutte le 24 configurazioni testate, mostrando affiancate le accuratezze di training e test per identificare immediatamente pattern di overfitting.  
La visualizzazione simultanea di train e test accuracy permette di valutare sia le prestazioni massime raggiungibili che la capacità di generalizzazione di ogni configurazione, elemento fondamentale per la selezione del modello ottimale.

In [ ]:
# Selezione migliori configurazioni per evidenziazione
migliore_mlp = max(risultati_mlp, key=lambda x: x['test_accuracy'])
migliore_cnn = max(risultati_cnn, key=lambda x: x['test_accuracy'])

# Preparazione dati per tutte le configurazioni
nomi_config = [r['nome_config'] for r in tutti_risultati]
acc_train_tutte = [r['train_accuracy'] for r in tutti_risultati]
acc_test_tutte = [r['test_accuracy'] for r in tutti_risultati]
tipi_modello = [r['tipo_modello'] for r in tutti_risultati]

# Separazione indici MLP e CNN
indici_mlp = [i for i, t in enumerate(tipi_modello) if t == 'MLP']
indici_cnn = [i for i, t in enumerate(tipi_modello) if t == 'CNN']

# Visualizzazione
fig, ax = plt.subplots(figsize=(16, 8))

# Posizioni delle barre
x = np.arange(len(nomi_config))
larghezza = 0.35

# Barre per accuratezza training e test
bars_train = ax.bar(x - larghezza/2, acc_train_tutte, larghezza, 
                   label='Accuratezza Training', alpha=0.8, color='lightcoral')
bars_test = ax.bar(x + larghezza/2, acc_test_tutte, larghezza, 
                  label='Accuratezza Test', alpha=0.8, color='steelblue')

# Colorazione diversa per MLP e CNN sui bordi
for i in indici_mlp:
    bars_train[i].set_edgecolor('darkred')
    bars_test[i].set_edgecolor('darkblue')
    bars_train[i].set_linewidth(1.5)
    bars_test[i].set_linewidth(1.5)

for i in indici_cnn:
    bars_train[i].set_edgecolor('orange')
    bars_test[i].set_edgecolor('green')
    bars_train[i].set_linewidth(2)
    bars_test[i].set_linewidth(2)

ax.set_xlabel('Configurazione')
ax.set_ylabel('Accuratezza')
ax.set_title('Confronto Completo: Accuratezza Training vs Test per Tutte le Configurazioni')
ax.set_xticks(x)
ax.set_xticklabels(nomi_config, rotation=45, ha='right')
ax.legend()
ax.grid(True, alpha=0.3)

# Evidenziazione delle migliori configurazioni
idx_migliore_mlp = tutti_risultati.index(migliore_mlp)
idx_migliore_cnn = tutti_risultati.index(migliore_cnn)

ax.annotate(f'Miglior MLP\n{migliore_mlp["test_accuracy"]:.4f}', 
           xy=(idx_migliore_mlp + larghezza/2, migliore_mlp['test_accuracy']),
           xytext=(10, 20), textcoords='offset points',
           bbox=dict(boxstyle='round,pad=0.3', facecolor='lightblue', alpha=0.7),
           arrowprops=dict(arrowstyle='->', color='blue'))

ax.annotate(f'Miglior CNN\n{migliore_cnn["test_accuracy"]:.4f}', 
           xy=(idx_migliore_cnn + larghezza/2, migliore_cnn['test_accuracy']),
           xytext=(10, -30), textcoords='offset points',
           bbox=dict(boxstyle='round,pad=0.3', facecolor='lightgreen', alpha=0.7),
           arrowprops=dict(arrowstyle='->', color='green'))

plt.tight_layout()
plt.show()

print("CONFRONTO ARCHITETTURE:")
print(f"Miglior MLP: {migliore_mlp['nome_config']} - Accuratezza Test: {migliore_mlp['test_accuracy']:.4f}")
print(f"Miglior CNN: {migliore_cnn['nome_config']} - Accuratezza Test: {migliore_cnn['test_accuracy']:.4f}")

#### Discussione del grafico e riflessione sui risultati
Il confronto completo rivela pattern distintivi tra MLP e CNN: le CNN mostrano consistentemente maggiore capacità di generalizzazione con gap train-test più contenuti (mediamente 0.0034-0.0114) rispetto agli MLP (0.0004-0.0201), indicando architetture intrinsecamente più robuste all'overfitting grazie ai meccanismi di condivisione dei pesi e alle operazioni di convoluzione che catturano invarianze spaziali.

La migliore configurazione **CNN** (*extended_lr0.001*: 98.82%) supera il miglior **MLP** (*250n_1S_lr0.001*: 98.10%) di 0.72 punti percentuali, dimostrando la superiorità delle architetture convoluzionali per dati visivi anche su dataset relativamente semplici come MNIST.

Particolarmente critico è l'effetto del learning rate 0.1 che causa collasso completo nelle CNN (accuratezza ~10%) suggerendo maggiore sensibilità all'instabilità di training, mentre gli MLP mostrano degrado graduale. 

Per applicazioni pratiche, si raccomanda l'uso di CNN con learning rate conservativi (≤0.01) quando le risorse computazionali lo permettono, riservando gli MLP a scenari con vincoli di velocità estremi dove la differenza di accuratezza è accettabile.

### Grafico 3: Analisi dell'Efficienza (Accuratezza per Secondo di Training)

Questo grafico quantifica l'efficienza di ogni configurazione calcolando il rapporto accuratezza/tempo, metrica fondamentale per applicazioni con vincoli temporali. L'efficienza rivela quale architettura offre il miglior ritorno in termini di prestazioni per unità di tempo investito, considerazione cruciale per deployment in produzione o sperimentazione rapida.

In [ ]:
# Calcolo efficienza per ogni configurazione
efficienze = [r['test_accuracy'] / r['training_time'] for r in tutti_risultati]
nomi_config_ordinati = []
efficienze_ordinate = []
tipi_ordinati = []

# Ordinamento per efficienza decrescente
indici_ordinati = sorted(range(len(efficienze)), key=lambda i: efficienze[i], reverse=True)

for i in indici_ordinati:
    nomi_config_ordinati.append(nomi_config[i])
    efficienze_ordinate.append(efficienze[i])
    tipi_ordinati.append(tipi_modello[i])

# Visualizzazione
fig, ax = plt.subplots(figsize=(16, 8))

# Colori diversi per MLP e CNN
colori = ['lightblue' if tipo == 'MLP' else 'salmon' for tipo in tipi_ordinati]
bordi = ['darkblue' if tipo == 'MLP' else 'darkred' for tipo in tipi_ordinati]

bars = ax.bar(range(len(nomi_config_ordinati)), efficienze_ordinate, 
              color=colori, edgecolor=bordi, linewidth=1.5, alpha=0.8)

ax.set_xlabel('Configurazione (ordinate per efficienza decrescente)')
ax.set_ylabel('Efficienza (Accuratezza / Tempo di Training)')
ax.set_title('Analisi Efficienza: Accuratezza per Secondo di Training')
ax.set_xticks(range(len(nomi_config_ordinati)))
ax.set_xticklabels(nomi_config_ordinati, rotation=45, ha='right')
ax.grid(True, alpha=0.3)

# Annotazioni per le configurazioni più efficienti
for i in range(min(5, len(bars))):  # Evidenzia top 5
    height = bars[i].get_height()
    ax.annotate(f'{height:.4f}', xy=(i, height),
               xytext=(0, 3), textcoords="offset points", 
               ha='center', va='bottom', fontweight='bold')

# Legenda manuale
ax.bar([], [], color='lightblue', alpha=0.8, label='MLP')
ax.bar([], [], color='salmon', alpha=0.8, label='CNN')
ax.legend()

plt.tight_layout()
plt.show()

# Calcolo statistiche di efficienza
eff_mlp = [efficienze[i] for i in range(len(tipi_modello)) if tipi_modello[i] == 'MLP']
eff_cnn = [efficienze[i] for i in range(len(tipi_modello)) if tipi_modello[i] == 'CNN']

print("ANALISI EFFICIENZA:")
print(f"Configurazione più efficiente: {nomi_config_ordinati[0]} - {efficienze_ordinate[0]:.4f} acc/s")
print(f"Efficienza media MLP: {np.mean(eff_mlp):.4f} acc/s")
print(f"Efficienza media CNN: {np.mean(eff_cnn):.4f} acc/s")
print(f"Rapporto efficienza MLP/CNN: {np.mean(eff_mlp)/np.mean(eff_cnn):.2f}x")

#### Discussione del grafico e riflessione sui risultati
L'analisi dell'efficienza rivela una dominanza netta degli **MLP** con le configurazioni più piccole che raggiungono i vertici della classifica (tipicamente >0.2 accuratezza/secondo), principalmente dovuta ai tempi di training drammaticamente inferiori (4.5-37.9s vs 48.3-112.6s delle CNN) che compensano ampiamente il leggero gap di accuratezza.  

Le **CNN**, nonostante prestazioni superiori, mostrano efficienza significativamente ridotta (media 0.018 vs 0.084 acc/s degli MLP) rappresentando un rapporto di 4.7x a favore degli MLP, confermando il trade-off fondamentale velocità-accuratezza nel machine learning. 
Le configurazioni MLP con learning rate moderati (0.01-0.001) e architetture snelle (50-100 neuroni, 1 strato) emergono come ideali per prototipazione rapida e deployment con vincoli temporali stretti. 

Dal punto di vista pratico, la scelta dovrebbe basarsi sui requisiti specifici: 
- MLP per iterazione veloce di sviluppo, validazione di proof-of-concept e sistemi real-time
- CNN quando l'accuratezza marginale giustifica l'investimento computazionale aggiuntivo, tipicamente in sistemi di produzione critici dove ogni frazione di punto percentuale ha valore economico.

### Grafico 4: Overfitting per Complessità del Modello

Questo grafico analizza la relazione tra complessità del modello (numero totale di parametri) e il fenomeno dell'overfitting (differenza tra accuratezza di training e test). La visualizzazione ordinata per complessità crescente permette di identificare soglie critiche oltre le quali i modelli iniziano a memorizzare anziché generalizzare, informazione cruciale per la progettazione di architetture bilanciate.

In [ ]:
# Preparazione dati ordinati per complessità
complessita = [r['parametri_totali'] for r in tutti_risultati]
overfitting_valori = [r['overfitting'] for r in tutti_risultati]

# Ordinamento per complessità crescente
indici_complessita = sorted(range(len(complessita)), key=lambda i: complessita[i])

nomi_ordinati_complessita = [nomi_config[i] for i in indici_complessita]
complessita_ordinata = [complessita[i] for i in indici_complessita]
overfitting_ordinato = [overfitting_valori[i] for i in indici_complessita]
tipi_ordinati_complessita = [tipi_modello[i] for i in indici_complessita]

# Visualizzazione
fig, ax = plt.subplots(figsize=(16, 8))

# Colori basati sul livello di overfitting
colori_overfitting = []
for ovf in overfitting_ordinato:
    if ovf < 0.01:
        colori_overfitting.append('lightgreen')  # Basso overfitting
    elif ovf < 0.02:
        colori_overfitting.append('gold')       # Moderato overfitting
    else:
        colori_overfitting.append('lightcoral') # Alto overfitting

bars = ax.bar(range(len(nomi_ordinati_complessita)), overfitting_ordinato, 
              color=colori_overfitting, alpha=0.8)

# Bordi diversi per tipo di modello
for i, tipo in enumerate(tipi_ordinati_complessita):
    if tipo == 'MLP':
        bars[i].set_edgecolor('darkblue')
        bars[i].set_linewidth(1.5)
    else:
        bars[i].set_edgecolor('darkred')
        bars[i].set_linewidth(2)

ax.set_xlabel('Configurazione (ordinate per complessità crescente)')
ax.set_ylabel('Overfitting (Accuratezza Training - Test)')
ax.set_title('Overfitting vs Complessità del Modello')
ax.set_xticks(range(len(nomi_ordinati_complessita)))
ax.set_xticklabels(nomi_ordinati_complessita, rotation=45, ha='right')
ax.grid(True, alpha=0.3)
ax.axhline(y=0, color='black', linestyle='--', alpha=0.5)

# Annotazioni per i valori più alti
soglia_annotazione = sorted(overfitting_ordinato, reverse=True)[4]  # Top 5
for i, (ovf, comp) in enumerate(zip(overfitting_ordinato, complessita_ordinata)):
    if ovf >= soglia_annotazione:
        ax.annotate(f'{ovf:.3f}\n{comp/1000:.0f}K param', 
                   xy=(i, ovf), xytext=(0, 10), 
                   textcoords='offset points', ha='center', fontsize=9)

# Legenda
from matplotlib.patches import Patch
legenda_elementi = [
    Patch(facecolor='lightgreen', label='Basso Overfitting (<0.01)'),
    Patch(facecolor='gold', label='Moderato Overfitting (0.01-0.02)'),
    Patch(facecolor='lightcoral', label='Alto Overfitting (>0.02)'),
    Patch(facecolor='white', edgecolor='darkblue', label='MLP'),
    Patch(facecolor='white', edgecolor='darkred', label='CNN')
]
ax.legend(handles=legenda_elementi, loc='upper left')

plt.tight_layout()
plt.show()

print("ANALISI OVERFITTING:")
print(f"Range overfitting MLP: {min([r['overfitting'] for r in risultati_mlp]):.4f} - {max([r['overfitting'] for r in risultati_mlp]):.4f}")
print(f"Range overfitting CNN: {min([r['overfitting'] for r in risultati_cnn]):.4f} - {max([r['overfitting'] for r in risultati_cnn]):.4f}")
print(f"Modello più complesso: {max(complessita_ordinata)/1000:.0f}K parametri")
print(f"Modello meno complesso: {min(complessita_ordinata)/1000:.0f}K parametri")

#### Discussione del grafico e riflessione sui risultati

L'analisi dell'overfitting rivela comportamenti distintivi tra architetture: 
- Le CNN mantengono controllo superiore dell'overfitting (range 0.0012-0.0114) anche con alta complessità parametrica grazie ai meccanismi intrinsechi di regolarizzazione (weight sharing, invarianze spaziali)
- Gli MLP mostrano variabilità maggiore (0.0004-0.0201) con particolare vulnerabilità nelle configurazioni più profonde e con learning rate sub-ottimali. 
  
Controintuitivamente, non emerge una correlazione diretta tra numero di parametri e overfitting, suggerendo che l'architettura e l'algoritmo di ottimizzazione sono più determinanti della mera complessità parametrica: le CNN con 260K parametri mostrano overfitting inferiore a MLP con 80K parametri. 

I modelli con learning rate 0.1 presentano pattern anomali (overfitting estremamente basso) dovuti al collasso del training piuttosto che a buona generalizzazione. 

Dal punto di vista pratico, l'early stopping si rivela efficace nel prevenire overfitting severo in entrambe le architetture, mentre la scelta dell'architettura (CNN vs MLP) e del learning rate hanno impatto più significativo della complessità assoluta, supportando approcci di progettazione che privilegiano l'appropriatezza architettonica rispetto alla semplice limitazione parametrica.

### Grafico 5: Velocità di Convergenza (Iterazioni per Configurazione)

Questo grafico confronta il numero di iterazioni necessarie per raggiungere la convergenza across tutte le configurazioni, rivelando l'efficienza algoritmica di diverse architetture e iperparametri. La velocità di convergenza è cruciale per la comprensione dell'ottimizzazione: configurazioni che convergono rapidamente indicano paesaggi di loss più favorevoli e gradient flow più efficace.

In [ ]:
# Preparazione dati convergenza
iterazioni_tutte = [r['iterations'] for r in tutti_risultati]

# Ordinamento per numero di iterazioni crescente
indici_iter = sorted(range(len(iterazioni_tutte)), key=lambda i: iterazioni_tutte[i])

nomi_ordinati_iter = [nomi_config[i] for i in indici_iter]
iterazioni_ordinate = [iterazioni_tutte[i] for i in indici_iter]
tipi_ordinati_iter = [tipi_modello[i] for i in indici_iter]
lr_ordinati = [tutti_risultati[i]['learning_rate'] for i in indici_iter]

# Visualizzazione
fig, ax = plt.subplots(figsize=(16, 8))

# Colori basati su learning rate
colori_lr = []
for lr in lr_ordinati:
    if lr == 0.001:
        colori_lr.append('lightgreen')
    elif lr == 0.01:
        colori_lr.append('gold')
    else:
        colori_lr.append('lightcoral')

bars = ax.bar(range(len(nomi_ordinati_iter)), iterazioni_ordinate, 
              color=colori_lr, alpha=0.8)

# Bordi per tipo di modello
for i, tipo in enumerate(tipi_ordinati_iter):
    if tipo == 'MLP':
        bars[i].set_edgecolor('darkblue')
        bars[i].set_linewidth(1.5)
    else:
        bars[i].set_edgecolor('darkred')
        bars[i].set_linewidth(2)

ax.set_xlabel('Configurazione (ordinate per iterazioni crescenti)')
ax.set_ylabel('Iterazioni per Convergenza')
ax.set_title('Velocità di Convergenza per Tutte le Configurazioni')
ax.set_xticks(range(len(nomi_ordinati_iter)))
ax.set_xticklabels(nomi_ordinati_iter, rotation=45, ha='right')
ax.grid(True, alpha=0.3)

# Annotazioni per valori significativi
for i in range(0, len(bars), 4):  # Ogni 4 configurazioni
    height = bars[i].get_height()
    ax.annotate(f'{int(height)}', xy=(i, height),
               xytext=(0, 3), textcoords="offset points", 
               ha='center', va='bottom', fontsize=9)

# Legenda
legenda_elementi = [
    Patch(facecolor='lightgreen', label='LR = 0.001'),
    Patch(facecolor='gold', label='LR = 0.01'),
    Patch(facecolor='lightcoral', label='LR = 0.1'),
    Patch(facecolor='white', edgecolor='darkblue', label='MLP'),
    Patch(facecolor='white', edgecolor='darkred', label='CNN')
]
ax.legend(handles=legenda_elementi)

plt.tight_layout()
plt.show()

# Statistiche per tipo di modello
iter_mlp = [r['iterations'] for r in risultati_mlp]
iter_cnn = [r['iterations'] for r in risultati_cnn]

print("ANALISI VELOCITÀ CONVERGENZA:")
print(f"Iterazioni medie MLP: {np.mean(iter_mlp):.1f}")
print(f"Iterazioni medie CNN: {np.mean(iter_cnn):.1f}")
print(f"Configurazione più veloce: {nomi_ordinati_iter[0]} - {iterazioni_ordinate[0]} iterazioni")
print(f"Configurazione più lenta: {nomi_ordinati_iter[-1]} - {iterazioni_ordinate[-1]} iterazioni")


#### Discussione del grafico e dei risultati  
La velocità di convergenza mostra pattern chiari legati all'architettura e agli iperparametri: le CNN convergono sistematicamente più velocemente (media 6.7 iterazioni) rispetto agli MLP (media 22.2 iterazioni) grazie a gradient flow più efficace e paesaggi di loss più regolari derivanti dalla struttura convoluzionale.  

Il learning rate gioca un ruolo determinante con LR=0.1 che causa convergenza prematura in configurazioni degradate e LR conservativi (0.001) che richiedono più iterazioni ma raggiungono soluzioni superiori.  
Le configurazioni CNN con LR elevati mostrano convergenza artificialmente rapida (6 iterazioni) dovuta al collasso del training piuttosto che a ottimizzazione efficace, mentre configurazioni MLP complesse con LR moderati richiedono fino a 43 iterazioni riflettendo la maggiore difficoltà di navigazione in spazi parametrici ad alta dimensionalità.  

Dal punto di vista dell'efficienza computazionale, nonostante la convergenza più lenta degli MLP in termini di epoche, il tempo totale rimane competitivo per le architetture snelle grazie al costo computazionale per iterazione significativamente inferiore.  

Questa analisi suggerisce che per applicazioni con budget computazionale limitato, MLP con architetture moderate (50-100 neuroni) e LR=0.01 offrono il miglior compromesso convergenza-prestazioni, mentre CNN giustificano il maggior costo iterativo quando l'accuratezza finale è prioritaria.

### Grafico 6: Effetto Scaling MLP (1 vs 2 Strati Nascosti)

Questo grafico analizza sistematicamente l'effetto della profondità nelle reti MLP confrontando prestazioni e tempi di training tra architetture a 1 e 2 strati nascosti. L'analisi rivela il trade-off fondamentale tra capacità espressiva (profondità) e efficienza computazionale, fornendo insights cruciali per la progettazione di architetture bilanciate su dataset di complessità moderata come MNIST.

In [ ]:
# Analisi scaling MLP
range_neuroni = neuroni_lista
acc_1_strato = []
acc_2_strati = []
tempo_1_strato = []
tempo_2_strati = []

for neuroni in range_neuroni:
    # 1 strato
    risultati_1s = [r for r in risultati_mlp if r['neuroni'] == neuroni and r['n_strati'] == 1]
    if risultati_1s:
        acc_1_strato.append(np.mean([r['test_accuracy'] for r in risultati_1s]))
        tempo_1_strato.append(np.mean([r['training_time'] for r in risultati_1s]))
    
    # 2 strati  
    risultati_2s = [r for r in risultati_mlp if r['neuroni'] == neuroni and r['n_strati'] == 2]
    if risultati_2s:
        acc_2_strati.append(np.mean([r['test_accuracy'] for r in risultati_2s]))
        tempo_2_strati.append(np.mean([r['training_time'] for r in risultati_2s]))

# Visualizzazione
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6))

# Subplot 1: Scaling accuratezza
ax1.plot(range_neuroni, acc_1_strato, 'o-', linewidth=2, markersize=8, 
         label='1 Strato Nascosto', color='blue')
ax1.plot(range_neuroni, acc_2_strati, 's-', linewidth=2, markersize=8, 
         label='2 Strati Nascosti', color='darkblue')

ax1.set_xlabel('Neuroni per Strato')
ax1.set_ylabel('Accuratezza Test')
ax1.set_title('Scaling MLP: Accuratezza vs Profondità')
ax1.legend()
ax1.grid(True, alpha=0.3)

# Annotazioni
for i, (neuroni, acc1, acc2) in enumerate(zip(range_neuroni, acc_1_strato, acc_2_strati)):
    ax1.annotate(f'{acc1:.3f}', (neuroni, acc1), textcoords="offset points", 
                xytext=(0,10), ha='center', color='blue', fontweight='bold')
    ax1.annotate(f'{acc2:.3f}', (neuroni, acc2), textcoords="offset points", 
                xytext=(0,-15), ha='center', color='darkblue', fontweight='bold')

# Subplot 2: Scaling tempo di training
ax2.plot(range_neuroni, tempo_1_strato, 'o-', linewidth=2, markersize=8, 
         label='1 Strato Nascosto', color='green')
ax2.plot(range_neuroni, tempo_2_strati, 's-', linewidth=2, markersize=8, 
         label='2 Strati Nascosti', color='darkgreen')

ax2.set_xlabel('Neuroni per Strato')
ax2.set_ylabel('Tempo di Training (secondi)')
ax2.set_title('Scaling MLP: Tempo di Training vs Profondità')
ax2.legend()
ax2.grid(True, alpha=0.3)

# Annotazioni tempo
for i, (neuroni, t1, t2) in enumerate(zip(range_neuroni, tempo_1_strato, tempo_2_strati)):
    ax2.annotate(f'{t1:.1f}s', (neuroni, t1), textcoords="offset points", 
                xytext=(0,10), ha='center', color='green', fontweight='bold')
    ax2.annotate(f'{t2:.1f}s', (neuroni, t2), textcoords="offset points", 
                xytext=(0,-15), ha='center', color='darkgreen', fontweight='bold')

plt.tight_layout()
plt.show()

print("ANALISI SCALING MLP:")
for i, neuroni in enumerate(range_neuroni):
    print(f"{neuroni} neuroni: 1S={acc_1_strato[i]:.4f} ({tempo_1_strato[i]:.1f}s), "
          f"2S={acc_2_strati[i]:.4f} ({tempo_2_strati[i]:.1f}s)")

# Calcolo differenze prestazioni
diff_acc = [acc_1_strato[i] - acc_2_strati[i] for i in range(len(range_neuroni))]
rapporto_tempo = [tempo_2_strati[i] / tempo_1_strato[i] for i in range(len(range_neuroni))]

print(f"\nDifferenza accuratezza media (1S - 2S): {np.mean(diff_acc):+.4f}")
print(f"Rapporto tempo medio (2S : 1S): {np.mean(rapporto_tempo):.2f}x")

#### Discussione dei grafici e riflessione sui risultati
L'analisi dello scaling rivela un risultato controintuitivo per MNIST: le architetture a 1 strato nascosto superano sistematicamente quelle a 2 strati con un vantaggio medio di +0.022 punti di accuratezza, suggerendo che la maggiore profondità introduce overfitting anziché migliorare l'espressività per dataset relativamente semplici come le cifre manoscritte. 

Il fenomeno è particolarmente evidente con architetture larghe (250 neuroni) dove il gap raggiunge 0.052 punti, indicando che l'aumento di parametri da profondità aggiuntiva eccede la complessità intrinseca del task causando memorizzazione del training set. 

Dal punto di vista computazionale, le architetture a 2 strati richiedono mediamente 1.11-1.68x più tempo per convergere, penalizzando ulteriormente il rapporto prestazioni-costo già sfavorevole. 
Questo comportamento riflette la natura del dataset MNIST dove le features discriminative sono relativamente semplici e non richiedono composizioni gerarchiche complesse che motiverebbero architetture profonde. 

Per applicazioni pratiche su MNIST, si raccomanda fortemente l'uso di architetture a singolo strato nascosto con 100-250 neuroni che offrono il miglior compromesso accuratezza-efficienza, riservando architetture più profonde a dataset con maggiore complessità strutturale dove i benefici della gerarchia di features giustifichino il costo computazionale aggiuntivo.

### Conclusioni

**Configurazioni ottimali identificate:**  

Gli esperimenti sistematici hanno identificato due architetture leader: 
- **MLP** con *250 neuroni, 1 strato nascosto* e *learning rate 0.001* raggiunge **98.10%** di accuratezza rappresentando la soluzione più efficiente, 
- **CNN** *extended* con *learning rate 0.001* ottiene **98.82%** stabilendo il nuovo benchmark di prestazioni con superiore robustezza all'overfitting. 

**Insights principali emergenti:**  

Il learning rate si conferma iperparametro critico con 0.001-0.01 come range ottimale, valori di 0.1 causano collasso catastrofico nelle CNN mentre rimangono tollerabili negli MLP.  

La profondità aggiuntiva negli MLP danneggia le prestazioni su MNIST introducendo overfitting senza benefici, contraddicendo l'intuizione comune sulla superiorità di architetture profonde.  

Le CNN mostrano intrinseca resistenza all'overfitting e convergenza più rapida ma richiedono 2.5-4x più tempo totale di training.

**Raccomandazioni strategiche:** 
- Per *prototipazione rapida e vincoli computazionali* utilizzare MLP(100, lr=0.01) che offre 97.3% accuratezza in <10 secondi
- Per *massimizzazione prestazioni senza vincoli temporali* impiegare CNN extended con lr=0.001 ottenendo 98.8% con robustezza superiore 
- Per *deployment critico* bilanciare con MLP(250, lr=0.001) che raggiunge 98.1% mantenendo efficienza 4x superiore alle CNN.

---
## Punto B: Analisi delle cifre più difficili da riconoscere

Utilizziamo la matrice di confusione per identificare quali cifre il modello MLP trova più difficili da classificare correttamente.

In [ ]:
# Addestro un MLP con architettura ottimale trovata precedentemente
best_mlp_config = max(mlp_results, key=lambda x: x['test_accuracy'])

mlp_best = MLPClassifier(
    hidden_layer_sizes=best_mlp_config['hidden_layers'],
    learning_rate_init=best_mlp_config['learning_rate'],
    max_iter=100,
    random_state=42,
    early_stopping=True,
    validation_fraction=0.1
)

print(f"Training MLP con architettura ottimale: {best_mlp_config['config_name']}")
mlp_best.fit(x_tr, mnist_tr_labels)
print(f"Accuratezza sul test set: {mlp_best.score(x_te, mnist_te_labels):.4f}")

NameError: name 'mlp_results' is not defined

In [ ]:
# Calcolo predizioni e matrice di confusione
y_pred = mlp_best.predict(x_te)

# Visualizzazione matrice di confusione
cm = metrics.confusion_matrix(mnist_te_labels, y_pred)
disp = metrics.ConfusionMatrixDisplay(confusion_matrix=cm)

fig, ax = plt.subplots(figsize=(10, 8))
disp.plot(ax=ax, cmap='Blues', values_format='d')
ax.set_title('Matrice di Confusione - MLP su MNIST', fontsize=16)
plt.show()

NameError: name 'mlp_best' is not defined

In [ ]:
# Analisi degli errori più frequenti
errors_per_digit = []
for digit in range(10):
    mask = mnist_te_labels == digit
    total = np.sum(mask)
    correct = np.sum((y_pred == mnist_te_labels) & mask)
    error_rate = 1 - (correct / total)
    
    errors_per_digit.append({
        'digit': digit,
        'total_samples': total,
        'correct': correct,
        'errors': total - correct,
        'error_rate': error_rate,
        'accuracy': correct / total
    })

df_errors = pd.DataFrame(errors_per_digit)
df_errors_sorted = df_errors.sort_values('error_rate', ascending=False)

print("Cifre ordinate per difficoltà (tasso di errore):")
print(df_errors_sorted[['digit', 'total_samples', 'errors', 'error_rate', 'accuracy']])

In [ ]:
# Visualizzazione delle coppie di cifre più confuse
confusion_pairs = []
for i in range(10):
    for j in range(10):
        if i != j and cm[i, j] > 0:
            confusion_pairs.append({
                'true_digit': i,
                'predicted_digit': j,
                'count': cm[i, j],
                'percentage': cm[i, j] / np.sum(cm[i, :]) * 100
            })

df_confusion = pd.DataFrame(confusion_pairs)
df_confusion_sorted = df_confusion.sort_values('count', ascending=False).head(10)

print("\nLe 10 coppie di cifre più confuse:")
for _, row in df_confusion_sorted.iterrows():
    print(f"{row['true_digit']} → {row['predicted_digit']}: {row['count']} errori ({row['percentage']:.1f}%)")

In [ ]:
# Visualizzazione esempi di cifre classificate erroneamente
fig, axes = plt.subplots(4, 5, figsize=(15, 12))
axes = axes.ravel()

example_idx = 0
for _, conf_pair in df_confusion_sorted.head(4).iterrows():
    true_digit = conf_pair['true_digit']
    pred_digit = conf_pair['predicted_digit']
    
    # Trovo esempi di questo tipo di errore
    error_mask = (mnist_te_labels == true_digit) & (y_pred == pred_digit)
    error_indices = np.where(error_mask)[0]
    
    # Mostro fino a 5 esempi per ogni coppia
    for i in range(min(5, len(error_indices))):
        if example_idx < 20:
            idx = error_indices[i]
            axes[example_idx].imshow(mnist_te_data[idx], cmap='gray')
            axes[example_idx].set_title(f'Vero: {true_digit}, Predetto: {pred_digit}', fontsize=10)
            axes[example_idx].axis('off')
            example_idx += 1

# Nascondo assi non utilizzati
for i in range(example_idx, 20):
    axes[i].axis('off')

plt.suptitle('Esempi di cifre classificate erroneamente', fontsize=16)
plt.tight_layout()
plt.show()

## Punto C: Curve psicometriche - Effetto del rumore

Seguendo la metodologia dell'articolo di Testolin et al. (2017), analizziamo come l'accuratezza degrada all'aumentare del rumore Gaussiano aggiunto alle immagini.

In [ ]:
# Funzione per aggiungere rumore Gaussiano
def add_gaussian_noise(images, noise_std):
    """
    Aggiunge rumore Gaussiano alle immagini.
    
    Args:
        images: array di immagini
        noise_std: deviazione standard del rumore
    
    Returns:
        Immagini con rumore, clippate tra 0 e 1
    """
    noise = np.random.normal(0, noise_std, images.shape)
    noisy_images = images + noise
    return np.clip(noisy_images, 0, 1)

In [ ]:
# Test con diversi livelli di rumore
noise_levels = np.arange(0, 0.5, 0.05)
accuracies_mlp = []

# Uso un subset del test set per velocizzare
subset_size = 2000
x_te_subset = x_te[:subset_size]
y_te_subset = mnist_te_labels[:subset_size]

print("Calcolo curve psicometriche per MLP...")
for noise_std in noise_levels:
    x_te_noisy = add_gaussian_noise(x_te_subset, noise_std)
    acc_mlp = mlp_best.score(x_te_noisy, y_te_subset)
    accuracies_mlp.append(acc_mlp)
    print(f"Noise std: {noise_std:.3f} - MLP acc: {acc_mlp:.4f}")

# Test anche con CNN se disponibile
best_cnn_config = max(cnn_results, key=lambda x: x['test_accuracy'])
cnn_model = create_cnn_model(best_cnn_config['architecture'], best_cnn_config['learning_rate'])

# Riaddestro il modello CNN migliore
early_stopping = keras.callbacks.EarlyStopping(patience=5, min_delta=0.001, restore_best_weights=True, verbose=0)
cnn_model.fit(x_tr_conv, mnist_tr_labels, validation_split=0.1, epochs=20, batch_size=128, 
              callbacks=[early_stopping], verbose=0)

print("\nCalcolo curve psicometriche per CNN...")
accuracies_cnn = []
x_te_conv_subset = x_te_conv[:subset_size]

for noise_std in noise_levels:
    x_te_conv_noisy = add_gaussian_noise(x_te_conv_subset, noise_std)
    test_loss, acc_cnn = cnn_model.evaluate(x_te_conv_noisy, y_te_subset, verbose=0)
    accuracies_cnn.append(acc_cnn)
    print(f"Noise std: {noise_std:.3f} - CNN acc: {acc_cnn:.4f}")

Calcolo curve psicometriche per MLP...


NameError: name 'mlp_best' is not defined

In [ ]:
# Visualizzazione curve psicometriche
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 6))

# Grafico 1: Curve psicometriche
ax1.plot(noise_levels, accuracies_mlp, 'o-', label='MLP', linewidth=3, markersize=8, color='blue')
ax1.plot(noise_levels, accuracies_cnn, 's-', label='CNN', linewidth=3, markersize=8, color='red')

ax1.set_xlabel('Deviazione standard del rumore', fontsize=12)
ax1.set_ylabel('Accuratezza', fontsize=12)
ax1.set_title('Curve Psicometriche - Robustezza al rumore', fontsize=14)
ax1.legend(fontsize=12)
ax1.grid(True, alpha=0.3)
ax1.set_ylim(0, 1.05)

# Evidenziare punti chiave
for i, (noise, acc_mlp, acc_cnn) in enumerate(zip(noise_levels[::2], accuracies_mlp[::2], accuracies_cnn[::2])):
    ax1.annotate(f'{acc_mlp:.2f}', (noise, acc_mlp), textcoords="offset points", 
                xytext=(0,10), ha='center', fontsize=9, color='blue')
    ax1.annotate(f'{acc_cnn:.2f}', (noise, acc_cnn), textcoords="offset points", 
                xytext=(0,-15), ha='center', fontsize=9, color='red')

# Grafico 2: Esempi di cifre con diversi livelli di rumore
noise_examples = [0, 0.1, 0.2, 0.3, 0.4]
digit_idx = 0

axes_noise = []
for i, noise in enumerate(noise_examples):
    ax_sub = fig.add_subplot(2, 5, 6+i)
    noisy_img = add_gaussian_noise(x_te[digit_idx:digit_idx+1], noise)[0]
    ax_sub.imshow(noisy_img.reshape(28, 28), cmap='gray', vmin=0, vmax=1)
    ax_sub.set_title(f'σ = {noise}', fontsize=10)
    ax_sub.axis('off')

plt.figtext(0.75, 0.02, f'Esempi di cifra {mnist_te_labels[digit_idx]} con diversi livelli di rumore', 
           ha='center', fontsize=12)

plt.tight_layout()
plt.show()

## Punto D: Effetto della riduzione dei dati di training

Analizziamo come le prestazioni degradano quando riduciamo drasticamente la quantità di dati di training disponibili.

In [ ]:
# Test con diverse percentuali di dati di training
train_percentages = [1, 5, 10, 25, 50, 75, 100]
results_data_reduction = []

print("Test con riduzione dei dati di training...")
for percentage in train_percentages:
    print(f"\nTraining con {percentage}% dei dati...")
    
    # Campionamento stratificato per mantenere bilanciamento classi
    indices = []
    for digit in range(10):
        digit_indices = np.where(mnist_tr_labels == digit)[0]
        n_digit_samples = int(len(digit_indices) * percentage / 100)
        if n_digit_samples > 0:
            selected_indices = np.random.choice(digit_indices, n_digit_samples, replace=False)
            indices.extend(selected_indices)
    
    indices = np.array(indices)
    x_tr_reduced = x_tr[indices]
    y_tr_reduced = mnist_tr_labels[indices]
    
    print(f"Samples utilizzati: {len(indices)} su {len(x_tr)}")
    
    # Training MLP
    mlp_reduced = MLPClassifier(
        hidden_layer_sizes=(100, 100),
        max_iter=100,
        random_state=42,
        early_stopping=True,
        validation_fraction=0.1 if len(indices) > 100 else 0.2
    )
    
    start_time = time.time()
    mlp_reduced.fit(x_tr_reduced, y_tr_reduced)
    training_time = time.time() - start_time
    
    train_acc = mlp_reduced.score(x_tr_reduced, y_tr_reduced)
    test_acc = mlp_reduced.score(x_te, mnist_te_labels)
    
    results_data_reduction.append({
        'percentage': percentage,
        'n_samples': len(indices),
        'train_accuracy': train_acc,
        'test_accuracy': test_acc,
        'overfitting': train_acc - test_acc,
        'training_time': training_time
    })
    
    print(f"Train acc: {train_acc:.4f}, Test acc: {test_acc:.4f}")

Test con riduzione dei dati di training...

Training con 1% dei dati...
Samples utilizzati: 596 / 60000
Train acc: 0.9664, Test acc: 0.8687

Training con 5% dei dati...
Samples utilizzati: 2996 / 60000
Train acc: 0.9907, Test acc: 0.9215

Training con 10% dei dati...
Samples utilizzati: 5996 / 60000
Train acc: 0.9887, Test acc: 0.9409

Training con 25% dei dati...
Samples utilizzati: 14995 / 60000
Train acc: 0.9944, Test acc: 0.9600

Training con 50% dei dati...
Samples utilizzati: 29997 / 60000
Train acc: 0.9976, Test acc: 0.9706

Training con 75% dei dati...
Samples utilizzati: 44995 / 60000
Train acc: 0.9975, Test acc: 0.9761

Training con 100% dei dati...
Samples utilizzati: 60000 / 60000
Train acc: 0.9972, Test acc: 0.9783


In [ ]:
# Visualizzazione effetto riduzione dati
df_reduction = pd.DataFrame(results_data_reduction)

fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(15, 10))

# Grafico 1: Accuratezza vs percentuale dati
ax1.plot(df_reduction['percentage'], df_reduction['test_accuracy'], 'o-', 
        linewidth=3, markersize=10, color='darkblue', label='Test')
ax1.plot(df_reduction['percentage'], df_reduction['train_accuracy'], 's-', 
        linewidth=3, markersize=10, color='lightblue', label='Train')
ax1.set_xlabel('Percentuale di dati di training utilizzati (%)')
ax1.set_ylabel('Accuratezza')
ax1.set_title('Effetto della riduzione dei dati di training')
ax1.legend()
ax1.grid(True, alpha=0.3)

# Evidenzio il punto al 10%
idx_10 = df_reduction[df_reduction['percentage'] == 10].index[0]
ax1.scatter(10, df_reduction.loc[idx_10, 'test_accuracy'], 
          s=200, color='red', zorder=5)
ax1.annotate(f"10%: {df_reduction.loc[idx_10, 'test_accuracy']:.3f}", 
           xy=(10, df_reduction.loc[idx_10, 'test_accuracy']),
           xytext=(20, df_reduction.loc[idx_10, 'test_accuracy'] - 0.05),
           arrowprops=dict(arrowstyle='->', color='red'),
           fontsize=11)

# Grafico 2: Overfitting vs dimensione dataset
ax2.plot(df_reduction['percentage'], df_reduction['overfitting'], 'o-', 
        linewidth=3, markersize=10, color='purple')
ax2.set_xlabel('Percentuale di dati (%)')
ax2.set_ylabel('Overfitting (Train - Test)')
ax2.set_title('Overfitting vs Dimensione dataset')
ax2.grid(True, alpha=0.3)
ax2.axhline(y=0, color='red', linestyle='--', alpha=0.5)

# Grafico 3: Tempo vs dimensione dataset
ax3.plot(df_reduction['n_samples'], df_reduction['training_time'], 'o-', 
        linewidth=3, markersize=10, color='green')
ax3.set_xlabel('Numero di campioni')
ax3.set_ylabel('Tempo di training (s)')
ax3.set_title('Tempo di training vs Dimensione dataset')
ax3.grid(True, alpha=0.3)

# Grafico 4: Efficienza (acc/tempo) vs dimensione
efficiency = df_reduction['test_accuracy'] / df_reduction['training_time']
ax4.plot(df_reduction['percentage'], efficiency, 'o-', 
        linewidth=3, markersize=10, color='orange')
ax4.set_xlabel('Percentuale di dati (%)')
ax4.set_ylabel('Efficienza (Accuratezza / Tempo)')
ax4.set_title('Efficienza vs Dimensione dataset')
ax4.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

## Punto E: Training con rumore per migliorare la robustezza

Verifichiamo se l'aggiunta di rumore durante il training può migliorare le prestazioni su dati di test rumorosi.

In [ ]:
# Training di modelli con diversi livelli di rumore nel training set
training_noise_levels = [0, 0.05, 0.1, 0.15, 0.2]
models_with_noise = {}

print("Training modelli con rumore nei dati di training...")
for train_noise in training_noise_levels:
    print(f"\nTraining con rumore std = {train_noise}")
    
    # Aggiungo rumore ai dati di training
    if train_noise > 0:
        x_tr_noisy = add_gaussian_noise(x_tr, train_noise)
    else:
        x_tr_noisy = x_tr
    
    # Training MLP
    mlp_noise = MLPClassifier(
        hidden_layer_sizes=(100, 100),
        max_iter=100,
        random_state=42,
        early_stopping=True,
        validation_fraction=0.1
    )
    
    start_time = time.time()
    mlp_noise.fit(x_tr_noisy, mnist_tr_labels)
    training_time = time.time() - start_time
    
    models_with_noise[train_noise] = mlp_noise
    
    # Test su dati puliti
    clean_acc = mlp_noise.score(x_te, mnist_te_labels)
    print(f"Accuratezza su test set pulito: {clean_acc:.4f}")
    print(f"Tempo di training: {training_time:.1f}s")

Training modelli con rumore nei dati di training...

Training con rumore std = 0
Accuratezza su test set pulito: 0.9803
Tempo di training: 21.3s

Training con rumore std = 0.05
Accuratezza su test set pulito: 0.9760
Tempo di training: 22.1s

Training con rumore std = 0.1
Accuratezza su test set pulito: 0.9688
Tempo di training: 10.5s

Training con rumore std = 0.15
Accuratezza su test set pulito: 0.9720
Tempo di training: 17.8s

Training con rumore std = 0.2
Accuratezza su test set pulito: 0.9622
Tempo di training: 15.7s


In [ ]:
# Test dei modelli su diversi livelli di rumore nel test set
test_noise_levels = np.arange(0, 0.4, 0.05)
results_noise_training = {}

print("\nTest dei modelli su dati rumorosi...")
for train_noise, model in models_with_noise.items():
    accuracies = []
    
    for test_noise in test_noise_levels:
        x_te_noisy = add_gaussian_noise(x_te_subset, test_noise)
        acc = model.score(x_te_noisy, y_te_subset)
        accuracies.append(acc)
    
    results_noise_training[train_noise] = accuracies
    print(f"Training noise {train_noise}: AUC = {np.trapz(accuracies, test_noise_levels):.3f}")


Test dei modelli su dati rumorosi...
Training noise 0: AUC = 0.290
Training noise 0.05: AUC = 0.279
Training noise 0.1: AUC = 0.313
Training noise 0.15: AUC = 0.326
Training noise 0.2: AUC = 0.330


In [ ]:
# Visualizzazione curve psicometriche con diversi livelli di rumore nel training
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(16, 6))

colors = plt.cm.viridis(np.linspace(0, 1, len(training_noise_levels)))

# Grafico 1: Curve psicometriche
for i, (train_noise, accuracies) in enumerate(results_noise_training.items()):
    ax1.plot(test_noise_levels, accuracies, 'o-', 
           label=f'Training noise σ = {train_noise}',
           color=colors[i], linewidth=2, markersize=6)

ax1.set_xlabel('Deviazione standard del rumore (test)', fontsize=12)
ax1.set_ylabel('Accuratezza', fontsize=12)
ax1.set_title('Effetto del rumore nel training sulla robustezza', fontsize=14)
ax1.legend(loc='upper right')
ax1.grid(True, alpha=0.3)
ax1.set_ylim(0, 1.05)

# Grafico 2: Analisi quantitativa del miglioramento
auc_scores = {}
for train_noise, accuracies in results_noise_training.items():
    auc = np.trapz(accuracies, test_noise_levels)
    auc_scores[train_noise] = auc

train_noises = list(auc_scores.keys())
aucs = list(auc_scores.values())

ax2.plot(train_noises, aucs, 'o-', linewidth=3, markersize=10, color='darkred')
ax2.set_xlabel('Rumore nel training (σ)', fontsize=12)
ax2.set_ylabel('AUC (Area Under Curve)', fontsize=12)
ax2.set_title('Area sotto la curva vs Rumore nel training', fontsize=14)
ax2.grid(True, alpha=0.3)

# Identifico il miglior livello
best_noise = max(auc_scores, key=auc_scores.get)
best_auc = auc_scores[best_noise]
ax2.scatter(best_noise, best_auc, s=200, color='gold', zorder=5)
ax2.annotate(f'Ottimo: σ={best_noise}\nAUC={best_auc:.3f}', 
           xy=(best_noise, best_auc),
           xytext=(best_noise + 0.05, best_auc - 0.5),
           arrowprops=dict(arrowstyle='->', color='gold'),
           fontsize=11, bbox=dict(boxstyle="round,pad=0.3", facecolor="white", alpha=0.8))

plt.tight_layout()
plt.show()

print(f"\nMiglior livello di rumore nel training: σ = {best_noise}")
print(f"Miglioramento rispetto al modello senza rumore: {(best_auc - auc_scores[0])/auc_scores[0]*100:.1f}%")

## Punto Bonus: Estensione con FashionMNIST

Replichiamo alcuni degli esperimenti precedenti utilizzando il dataset FashionMNIST, che presenta maggiore complessità.

In [ ]:
# Caricamento FashionMNIST
print("Caricamento FashionMNIST...")
fashion_tr = FashionMNIST(root="./data", train=True, download=True)
fashion_te = FashionMNIST(root="./data", train=False, download=True)

Caricamento FashionMNIST...


In [ ]:
# Preprocessing FashionMNIST
fashion_tr_data, fashion_tr_labels = fashion_tr.data.numpy(), fashion_tr.targets.numpy()
fashion_te_data, fashion_te_labels = fashion_te.data.numpy(), fashion_te.targets.numpy()

x_fashion_tr = fashion_tr_data.reshape(60000, 28 * 28) / 255.0
x_fashion_te = fashion_te_data.reshape(10000, 28 * 28) / 255.0

# Nomi delle classi
fashion_classes = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
                  'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

print(f"FashionMNIST caricato: {x_fashion_tr.shape[0]} train, {x_fashion_te.shape[0]} test")

FashionMNIST caricato: 60000 train, 10000 test


In [ ]:
# Visualizzazione esempi FashionMNIST
fig, axes = plt.subplots(2, 5, figsize=(15, 8))
axes = axes.ravel()

for i in range(10):
    idx = np.where(fashion_tr_labels == i)[0][0]
    axes[i].imshow(fashion_tr_data[idx], cmap='gray')
    axes[i].set_title(f'{i}: {fashion_classes[i]}', fontsize=12)
    axes[i].axis('off')

plt.suptitle('Esempi dal dataset FashionMNIST', fontsize=16)
plt.tight_layout()
plt.show()

In [ ]:
# Training MLP su FashionMNIST con stessa architettura ottimale
mlp_fashion = MLPClassifier(
    hidden_layer_sizes=best_mlp_config['hidden_layers'],
    learning_rate_init=best_mlp_config['learning_rate'],
    max_iter=100,
    random_state=42,
    early_stopping=True,
    validation_fraction=0.1
)

print(f"Training MLP su FashionMNIST con architettura: {best_mlp_config['config_name']}")
start_time = time.time()
mlp_fashion.fit(x_fashion_tr, fashion_tr_labels)
fashion_training_time = time.time() - start_time

fashion_train_acc = mlp_fashion.score(x_fashion_tr, fashion_tr_labels)
fashion_test_acc = mlp_fashion.score(x_fashion_te, fashion_te_labels)

print(f"Training time: {fashion_training_time:.1f}s")
print(f"Train accuracy: {fashion_train_acc:.4f}")
print(f"Test accuracy: {fashion_test_acc:.4f}")
print(f"Overfitting: {fashion_train_acc - fashion_test_acc:+.4f}")

# Confronto con MNIST
mnist_test_acc = mlp_best.score(x_te, mnist_te_labels)
print(f"\nConfronto con MNIST:")
print(f"MNIST test accuracy: {mnist_test_acc:.4f}")
print(f"FashionMNIST test accuracy: {fashion_test_acc:.4f}")
print(f"Differenza: {mnist_test_acc - fashion_test_acc:+.4f}")

NameError: name 'best_mlp_config' is not defined

In [ ]:
# Curve psicometriche comparative MNIST vs FashionMNIST
noise_levels_comp = np.arange(0, 0.3, 0.05)
acc_mnist = []
acc_fashion = []

# Subset per velocità
x_fashion_te_subset = x_fashion_te[:2000]
y_fashion_te_subset = fashion_te_labels[:2000]

print("Calcolo curve psicometriche comparative...")
for noise_std in noise_levels_comp:
    # MNIST
    x_noisy_mnist = add_gaussian_noise(x_te_subset, noise_std)
    acc_mnist.append(mlp_best.score(x_noisy_mnist, y_te_subset))
    
    # FashionMNIST
    x_noisy_fashion = add_gaussian_noise(x_fashion_te_subset, noise_std)
    acc_fashion.append(mlp_fashion.score(x_noisy_fashion, y_fashion_te_subset))
    
    print(f"Noise {noise_std:.2f}: MNIST {acc_mnist[-1]:.3f}, Fashion {acc_fashion[-1]:.3f}")

In [ ]:
# Visualizzazione comparativa finale
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(16, 12))

# Grafico 1: Curve psicometriche comparative
ax1.plot(noise_levels_comp, acc_mnist, 'o-', label='MNIST', 
         linewidth=3, markersize=8, color='blue')
ax1.plot(noise_levels_comp, acc_fashion, 's-', label='FashionMNIST', 
         linewidth=3, markersize=8, color='red')
ax1.set_xlabel('Deviazione standard del rumore', fontsize=12)
ax1.set_ylabel('Accuratezza', fontsize=12)
ax1.set_title('Confronto robustezza al rumore:\nMNIST vs FashionMNIST', fontsize=14)
ax1.legend(fontsize=12)
ax1.grid(True, alpha=0.3)
ax1.set_ylim(0, 1.05)

# Grafico 2: Matrice di confusione FashionMNIST
y_pred_fashion = mlp_fashion.predict(x_fashion_te)
cm_fashion = metrics.confusion_matrix(fashion_te_labels, y_pred_fashion)

im = ax2.imshow(cm_fashion, cmap='Blues')
ax2.set_xticks(range(10))
ax2.set_yticks(range(10))
ax2.set_xticklabels([f'{i}' for i in range(10)])
ax2.set_yticklabels([f'{i}: {fashion_classes[i][:7]}' for i in range(10)], fontsize=10)
ax2.set_xlabel('Predetto', fontsize=12)
ax2.set_ylabel('Vero', fontsize=12)
ax2.set_title('Matrice di Confusione\nFashionMNIST', fontsize=14)

# Grafico 3: Confronto accuratezze per classe
fashion_class_accs = []
mnist_class_accs = []

for digit in range(10):
    # FashionMNIST
    mask_f = fashion_te_labels == digit
    acc_f = np.sum((y_pred_fashion == fashion_te_labels) & mask_f) / np.sum(mask_f)
    fashion_class_accs.append(acc_f)
    
    # MNIST
    mask_m = mnist_te_labels == digit
    acc_m = np.sum((y_pred == mnist_te_labels) & mask_m) / np.sum(mask_m)
    mnist_class_accs.append(acc_m)

x_pos = np.arange(10)
width = 0.35

ax3.bar(x_pos - width/2, mnist_class_accs, width, label='MNIST', alpha=0.8, color='blue')
ax3.bar(x_pos + width/2, fashion_class_accs, width, label='FashionMNIST', alpha=0.8, color='red')
ax3.set_xlabel('Classe', fontsize=12)
ax3.set_ylabel('Accuratezza per classe', fontsize=12)
ax3.set_title('Accuratezza per classe:\nMNIST vs FashionMNIST', fontsize=14)
ax3.set_xticks(x_pos)
ax3.set_xticklabels([f'{i}' for i in range(10)])
ax3.legend()
ax3.grid(True, alpha=0.3)

# Grafico 4: Confronto errori più frequenti FashionMNIST
fashion_confusion_pairs = []
for i in range(10):
    for j in range(10):
        if i != j and cm_fashion[i, j] > 0:
            fashion_confusion_pairs.append({
                'true_class': fashion_classes[i],
                'pred_class': fashion_classes[j],
                'count': cm_fashion[i, j]
            })

df_fashion_confusion = pd.DataFrame(fashion_confusion_pairs)
top_fashion_errors = df_fashion_confusion.nlargest(8, 'count')

y_pos = np.arange(len(top_fashion_errors))
ax4.barh(y_pos, top_fashion_errors['count'], color='coral', alpha=0.8)
ax4.set_yticks(y_pos)
ax4.set_yticklabels([f"{row['true_class'][:6]} → {row['pred_class'][:6]}" 
                    for _, row in top_fashion_errors.iterrows()], fontsize=10)
ax4.set_xlabel('Numero di errori', fontsize=12)
ax4.set_title('Top 8 errori più frequenti\nFashionMNIST', fontsize=14)
ax4.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

## Conclusioni

### Riepilogo dei risultati principali:

[risultati da implementare]

1. **Effetto degli iperparametri (Punto A):**
   - [analisi basata sui risultati numerici]

2. **Cifre più difficili (Punto B):**
   - [analisi pattern errori specifici]

3. **Robustezza al rumore (Punto C):**
   - [confronto degradazione MLP vs CNN]

4. **Effetto dei dati di training (Punto D):**
   - [analisi prestazioni con dataset ridotto]

5. **Training con rumore (Punto E):**
   - [valutazione miglioramenti robustezza]

6. **FashionMNIST (Bonus):**
   - [confronto complessità dataset]

### Implicazioni pratiche:

[raccomandazioni basate sui risultati]

In [ ]:
# Statistiche finali del progetto
print("="*60)
print("RIEPILOGO FINALE DEL PROGETTO")
print("="*60)

print(f"\nPunto A - Analisi Iperparametri:")
print(f"  • Esperimenti MLP: {len(mlp_results)}")
print(f"  • Esperimenti CNN: {len(cnn_results)}")
print(f"  • Miglior MLP: {best_mlp_config['config_name']} -> Acc: {best_mlp_config['test_accuracy']:.4f}")
print(f"  • Miglior CNN: {best_cnn_config['config_name']} -> Acc: {best_cnn_config['test_accuracy']:.4f}")

print(f"\nPunto B - Analisi Errori:")
print(f"  • Cifra più difficile: {df_errors_sorted.iloc[0]['digit']} (Error rate: {df_errors_sorted.iloc[0]['error_rate']:.3f})")
print(f"  • Cifra più facile: {df_errors_sorted.iloc[-1]['digit']} (Error rate: {df_errors_sorted.iloc[-1]['error_rate']:.3f})")
print(f"  • Confusione più frequente: {df_confusion_sorted.iloc[0]['true_digit']} → {df_confusion_sorted.iloc[0]['predicted_digit']} ({df_confusion_sorted.iloc[0]['count']} errori)")

print(f"\nPunto C - Robustezza al Rumore:")
print(f"  • Livelli di rumore testati: {len(noise_levels)}")
print(f"  • Accuratezza senza rumore MLP: {accuracies_mlp[0]:.4f}")
print(f"  • Accuratezza senza rumore CNN: {accuracies_cnn[0]:.4f}")

print(f"\nPunto D - Riduzione Dati:")
print(f"  • Accuratezza con 100% dati: {df_reduction[df_reduction['percentage']==100]['test_accuracy'].iloc[0]:.4f}")
print(f"  • Accuratezza con 10% dati: {df_reduction[df_reduction['percentage']==10]['test_accuracy'].iloc[0]:.4f}")

print(f"\nPunto E - Training con Rumore:")
print(f"  • Livelli testati: {len(training_noise_levels)}")
print(f"  • Miglior configurazione: σ = {best_noise}")

print(f"\nBonus - FashionMNIST:")
print(f"  • Accuratezza MNIST: {mnist_test_acc:.4f}")
print(f"  • Accuratezza FashionMNIST: {fashion_test_acc:.4f}")

print(f"\n{'='*60}")
print("PROGETTO COMPLETATO CON SUCCESSO!")
print("Tutti i 5 punti + bonus implementati e analizzati.")
print("="*60)

RIEPILOGO FINALE DEL PROGETTO

Punto A - Analisi Iperparametri:


NameError: name 'mlp_results' is not defined